### PR TITLE
Fix and optimize `bin1d_vec` + extend units tests + fix imprecise bin edge creation

### LINKED ISSUE #255: Spatial binning doesn't properly account for floating point precision
I tested forecast models with a catalog whose events sometimes occur exactly on a cell boundary.
I found that [`region.get_index_of`](https://github.com/SCECcode/pycsep/blob/20be367d88bb81f64c8f3d5d433ffe856c6a9996/csep/core/regions.py#L618) generally does not assign them the spatial bin according to what is expected from

> _"inclusive on the lower bound and exclusive on the upper bound."_ [[`utils.calc.bin1d_vec` docstring](https://github.com/SCECcode/pycsep/blob/20be367d88bb81f64c8f3d5d433ffe856c6a9996/csep/utils/calc.py#L54)]

but the opposite (i.e., the neighboring bin). Although this function considers a tolerance (`numpy.finfo(numpy.float64).eps`) to (supposedly) account for numerical precision, I found that it is not sufficient to get the desired binning behavior, even after the revision in [commit 949d342c](https://github.com/SCECcode/pycsep/commit/949d342c181a984071e0f8e83a7460e631bb39c2).

I believe the corresponding [unit tests](https://github.com/SCECcode/pycsep/blob/20be367d88bb81f64c8f3d5d433ffe856c6a9996/tests/test_calc.py#L29) don't catch these literal "corner cases", as those tests are too constructed/synthetic and not based on real data such as a loaded CSEP region. Specifically, they do not consider issues due to _double_ precision (float64).

### 1. The problem

Let's do a quick self check:

```python
region = csep.core.regions.california_relm_region()

idx = 3
origin = region.origins()[idx]
print(idx, origin)

idx2 = region.get_index_of(*origin)
print(idx2, region.origins()[idx2])  # not the same as `idx` and `origin`
```

outputs:
```
3 [-125.4   40.4]
2 [-125.4   40.3]
```

Ooops. Not even a bin's origin is considered to be inside its own bin (but the neighboring one)!

> Btw: the above output may represent the coordinates as more precise than they numerically are:
> ```
> print(idx, [x for x in origin])
> ```
> gives: 
> ```
> 3 [-125.39999999999999, 40.400000000000006]
> ```
> which highlights that our unit tests are currently too optimistic/simplistic.

### 2. The correction

Only when adding a tiny amount to the Latitude and Longitude of the event coordinates gives the desired behavior:

```python
tinyamount = 1e-12

idx3 = region.get_index_of(
    origin[0] + tinyamount,
    origin[1] + tinyamount
)
print(idx3, region.origins()[idx3])  # same as `idx` and `origin`
```

outputs:
```
3 [-125.4   40.4]
```

Now we can do this self check for _all_ origins in the region:

```python
# tinyamount = 0  # uncomment for raising an Error
for i, origin in enumerate(region.origins()):

    assert i == region.get_index_of(
        origin[0] + tinyamount,
        origin[1] + tinyamount
    )
```

Only if (at least) `tinyamount = 1e-12` is added, the loop runs without Error. 

> Side note: Actually, if `tinyamount = 0`, it wouldn't raise an `AssertionError`, but this [`ValueError`](https://github.com/SCECcode/pycsep/blob/20be367d88bb81f64c8f3d5d433ffe856c6a9996/csep/core/regions.py#L633) in `get_index_of` for the first bin, because `get_index_of` tries to assign a bin that lies East of it (which doesn't exist and is outside the region). You can get an `AssertionError` like so:
> ```python
> tinyamount = 0
> lons = set()
> for i, origin in enumerate(region.origins()):
>
>     if origin[0] not in lons:
>         lons.add(origin[0])
>         continue
>     
>     assert i == region.get_index_of(
>         origin[0] + tinyamount,
>         origin[1] + tinyamount
>     )
> ```

### 3. The complexity

But this numerical issue gets slightly more complicated than that: 

#### a) Influence of the region

The region influences (i) the `tinyamount` that is just sufficient to bin correctly, and (ii) which coordinate component(s) must be corrected. For instance for Italy, this is sufficient:

```python
region = csep.core.regions.italy_csep_region()
tinyamount = 2e-13  # Note #1: smaller than for California region
for i, origin in enumerate(region.origins()):

    assert i == region.get_index_of(
        origin[0],  # Note #2: adding `tinyamount` to the Longitude NOT needed (!)
        origin[1] + tinyamount
    )
```

I don't know why, but it's certainly related to numerical representations of values.

#### b) Influence of the floating-point format/precision

The just-sufficient `tinyamount` also depends on the floating-point format (i.e., precision). For instance, assuming the events have single precision (float32), this is necessary:

```python
tinyamount = 1.6e-6  # for Italy region
tinyamount = 3.1e-6  # for California region
for i, origin in enumerate(region.origins()):

    assert i == region.get_index_of(
        origin[0].astype(np.float32) + tinyamount,  # (adding `tinyamount` not necessary for Italy region)
        origin[1].astype(np.float32) + tinyamount
    )
```

#### Summary

I tried to find out the just-sufficient threshold also as function of the corresponding `.eps` and `.resolution` property of `np.finfo` for the used precision (e.g., `np.finfo(np.float64).eps`):

| Region     | Precision | `tinyamount` | `.resolution *` | `.eps *` |
|------------|-----------|--------------|----------------:|---------:|
| California | float64   | 1.0e-12      |          1000.0 |  ~4400.0 |
|    Italy   | float64   | 1.4e-13      |           140.0 |   ~630.0 |
| California | float32   | 3.1e-6       |             3.1 |    ~26.0 |
|    Italy   | float32   | 1.6e-6       |             1.6 |    ~13.0 |

Apparently, `tinyamount` not only depends on the region, but it also doesn't scale with the precision.

## Suggested solution(s)

Some options (with increasing personal preference):
 1. add the required `tinyamount` to event coordinates;
 2. use `bin1d_vec`'s `tol` argument, which is not used by the `get_index_of` wrapper;
 3. revise the [tolerance section of `bin1d_vec`](https://github.com/SCECcode/pycsep/blob/20be367d88bb81f64c8f3d5d433ffe856c6a9996/csep/utils/calc.py#L82-L84);
 4. simplify the tolerance section of `bin1d_vec` by only adding a `tinyamount`.

**Option 1** is only a temporary (quick-)fix.

**For Option 2**, note that `tol = tinyamount` is not sufficient due to the [follow-up 'absolute tolerance' section](https://github.com/SCECcode/pycsep/blob/20be367d88bb81f64c8f3d5d433ffe856c6a9996/csep/utils/calc.py#L87-L90C). It needs to be slightly higher, e.g., `tol = tinyamount * 1.1` in our cases.

**For Option 3**, we _could_ do:

```diff
-    a0_tol = numpy.abs(a0) * numpy.finfo(numpy.float64).eps
-    h_tol = numpy.abs(h) * numpy.finfo(numpy.float64).eps
-    p_tol = numpy.abs(p) * numpy.finfo(numpy.float64).eps
+#    tol_fact = numpy.finfo(numpy.float64).eps  # (current)
+    tol_fact = # NEW THRESHOLD NEEDED
+    a0_tol = numpy.abs(a0) * tol_fact
+    h_tol = numpy.abs(h) * tol_fact
+    p_tol = numpy.abs(p) * tol_fact
```

Compared to `tol`, `tol_fact` is relative; I tried to find out the necessary thresholds again:

| Region     | Precision | `tol_fact` | `.resolution *` | `.eps *` |
|------------|-----------|------------|----------------:|---------:|
| California | float64   | 4.14e-15   |            4.14 |     18.7 |
|    Italy   | float64   | 1.74e-15   |            1.74 |      7.9 |
| California | float32   | 2.38e-8    |          0.0238 |      0.2 |
|    Italy   | float32   | 2.35e-8    |          0.0235 |      0.2 |

It also doesn't scale with the precision of the floating-point format, so I don't see the benefit of this option.

**For Option 4**, we could revert the [whole current tolerance section](https://github.com/SCECcode/pycsep/blob/20be367d88bb81f64c8f3d5d433ffe856c6a9996/csep/utils/calc.py#L82C1-L93C75) that was introduced in [commit 949d342c](https://github.com/SCECcode/pycsep/commit/949d342c181a984071e0f8e83a7460e631bb39c2) to a [prior state](https://github.com/SCECcode/pycsep/blob/0b1feb94a7aa078f1f21f0994208e99449d31d77/csep/utils/calc.py#L74C1-L77C42) and extend it (keeping the `tol` argument):

```diff
-    a0_tol = numpy.abs(a0) * numpy.finfo(numpy.float64).eps
-    h_tol = numpy.abs(h) * numpy.finfo(numpy.float64).eps
-    p_tol = numpy.abs(p) * numpy.finfo(numpy.float64).eps
-
-
-    # absolute tolerance
-    if tol is None:
-        idx = numpy.floor((p + (p_tol + a0_tol) - a0) / (h - h_tol))
-    else:
-        idx = numpy.floor((p + (tol + a0_tol) - a0) / (h - h_tol))
+    tol = tol or 0
+
+    dtype = np.asarray(p).dtype
+    if dtype.name in ('float64', 'float32'):
+        float_tol = numpy.finfo(dtype).resolution * {
+            'float64': 2000,
+            'float32': 10,
+        }[dtype.name]
+        tol = max(tol, float_tol)
+
+    idx = numpy.floor((p + tol - a0) / h)
```

I added a safety factor of 2-3 compared to Table 1 just to be more on the safe side.

Btw: We could even _always_ add `numpy.finfo(numpy.float32).resolution * 10 = 1e-5` independent of the floating-point format, which is on the geographical order of ~1 meter. So instead of the above change, this one:
```diff
+    float_min_tol = 1e-5  # conservative (minimum) tolerance
+    tol = max(tol or 0, float_min_tol)
+    idx = numpy.floor((p + tol - a0) / h)
```
This would be the same as (and replace!) setting/fixing `tol=0.00001` when binning the magnitude [across various places in pyCSEP code](https://github.com/search?q=repo%3ASCECcode%2Fpycsep+tol%3D0.0000&type=code).

### The implication

_If a catalog contains such "corner case" events_, fixing this issue _will_ change the test statistic – but only slightly.
Nevertheless, it would lead to **irreproducible previous results** (e.g., in my case for IGPE, ~at the 3 to 5th decimal digit).
Exactly such slight difference (compared to an independent binning implementation) was the reason for spotting this issue.

<br>

---

**A personal note**

Considering that event locations are sometimes very rough, it might have been a better idea to define grid cells so that their centers lie on rough coordinates (e.g., 13.5°), rather than their boundaries (13.45° / 13.55°); i.e., shift the grid by half a bin size in each direction.

---

**P.S.**

@khawajasim: Could [`QuadtreeGrid2D._find_location`](https://github.com/SCECcode/pycsep/blob/20be367d88bb81f64c8f3d5d433ffe856c6a9996/csep/core/regions.py#L1087) be affected by the same issue?

### PR DESCRIPTION
**Pull request by mherrmann3**
⚠ Caution, longer treatise ahead! 🤷‍♂️

### 1. Solving numerical issues (#255)

I [previously proposed a preliminary fix](https://github.com/SCECcode/pycsep/issues/255#issuecomment-2548687127) based on rounding the estimated bin size `h`. But there's a more elegant [fix](https://github.com/SCECcode/pycsep/commit/e12f6a0):

```diff
-    h_tol = numpy.abs(h) * numpy.finfo(numpy.float64).eps
+    h_tol = a0_tol
```
in [this line](https://github.com/SCECcode/pycsep/blob/2feb9148/csep/utils/calc.py#L83).

This change is sufficient because the prior floating-point operation `h = bins[1] - bins[0]` in the [previous line](https://github.com/SCECcode/pycsep/blob/2feb9148/csep/utils/calc.py#L83) carries the unit [roundoff error](https://en.wikipedia.org/wiki/Round-off_error) of the involved numbers (and may introduce a further roundoff error). To account for it, the tolerance must be based on machine `eps`ilon and the _involved_ numbers (not the resulting number). In our case, the two involved numbers in calculating `h` are of the same order of magnitude, and one of it is `a0` (i.e., the first bin edge `bin[0] == min(bins)`), for which we already calculate the tolerance (`a0_tol`). So we can use `a0_tol` also as tolerance for `h`.

Here's an illustration:

```
       bin1:   31.50000000000000000 (a0)
       bin2:   31.60000000000000142

          h:    0.10000000000000142 (bin2 - a0)
imprecision:    0.00000000000000142 (|h - 0.1|)

        eps:    0.00000000000000022 (FYI)
    |h|*eps:    0.00000000000000002 (previous h_tol)

   |a0|*eps:    0.00000000000000699 (new h_tol)
```

To account for the floating-point/roundoff errors, `h_tol` must be larger than the `imprecision` of `h`.
Only the new `h_tol` estimation guarantees that by reusing `a0_tol = abs(a0) * eps`.

> FYI: accounting for `h_tol` would not be necessary if the actual bin size `h` is accurately specified (e.g., `0.1`), i.e., not based on a prior operation. `CartesianGrid` already stores it as meta-data `.dh`, which is _exactly_ `0.1` for all currently implemented regions. `bin1d_vec` _could_ be adapted to receive this bin size as an additional argument. But `bin1d_vec` is also [used for binning magnitudes](https://github.com/search?q=repo%3ASCECcode%2Fpycsep+bin1d_vec%28&type=code), for which the bin size is not explicitly stored in the `region`. Of course we could change that and pass it to `bin1d_vec`, but I believe it's not necessary (for now).

#### 2.1 to `bin1d_vec`

Working with and looking at `bin1d_vec` way too often, I noticed that it can be simplified, made more robust/versatile _and_ (slightly) speed up by addressing various aspects. In the following, I list each incremental enhancement and measure its performance; I committed them separately for convenience and providing related notes in their description.

To quantify the performance speed-up, I timed the self-check over all origins for California, but – to avoid the overhead in `get_index_of` – only for one coordinate component by directly calling `bin1d_vec` in two ways:
1. looped (7682 individual call): `%timeit -r 300 -n 1 for origin in region.origins(): bin1d_vec(origin[0], region.xs)`
2. vectorized (vector with 7682 values), which matters much more in practice: `%timeit -r 6000 -n 100 bin1d_vec(origins_lon, region.xs)`

| Enhancement                                             | **vectorized** |  looped |
|:--------------------------------------------------------|---------------:|--------:|
| [Only the bare fix](https://github.com/SCECcode/pycsep/commit/e12f6a0)                                       |    **23.8 µs** | 36.3 ms |
| [1. Reorganize + make tolerance depend on _dtype_](https://github.com/SCECcode/pycsep/commit/79c08dd)        |    **24.1 µs** | 37.7 ms |
| [2. Replace `.min(bins)` with `bins[0]`](https://github.com/SCECcode/pycsep/commit/ec74058)                  |    **22.2 µs** | 26.3 ms |
| [3. Use `.asarray()` instead of `.array()`](https://github.com/SCECcode/pycsep/commit/22aac12)               |    **20.7 µs** | 24.8 ms |
| [4. Avoid try/except by always operating on arrays](https://github.com/SCECcode/pycsep/commit/1fac3b)       |    **20.8 µs** | 31.4 ms |

<br>

> Notes about `bin1d_vec` (for those interested):
> * As mentioned in `bin1d_vec`'s docstring, ["`bins [...] must be monotonically increasing`"](https://github.com/SCECcode/pycsep/blob/2feb9148/csep/utils/calc.py#L63) because the implemented `idx` calculation assumes/requires sorted bins (for this reason, there is no need for `.min()`). This is not checked, presumably because _pyCSEP_ itself never passes unsorted bins to this method (they always originate from ranges). Nevertheless, I added a check at the top, but commented it out; we may activate it in the future if there is need - for now, it only serves as an additional hint.
> * for passing the unit tests mentioned below, the necessary tolerances to add in the `idx` calculation depend on the _dtype_ of `p`:
>     * _float64_ doesn't need adding `a0_tol` nor `p_tol`;
>     * _float32_ doesn't need adding `a0_tol`, but `p_tol` (since `p` is _float32_);
>     * --> So `a0_tol` could be omitted, but to avoid unforeseeable future oddities, let's keep it anyway (and because it theoretically makes sense adding it).
> * the grouping and order of terms in the `idx` calculation influence the roundoff error, see [_What Every Computer Scientist Should Know About Floating-Point Arithmetic_](https://docs.oracle.com/cd/E19957-01/806-3568/ncg_goldberg.html). Illustration:
> ```
>  40.40000000000000568 (p)
>  31.50000000000000000 (a0)
>   0.00000000000003728 (p_tol)
>   0.00000000000003679 (a0_tol)
>
>   8.90000000000000568 (p - a0)
>   0.00000000000007407 (p_tol + a0_tol)
>
>   8.90000000000007674  <--  p + (p_tol + a0_tol) - a0  [ORIGINAL]
>   8.90000000000007674  <--  p + p_tol + a0_tol - a0
>   8.90000000000008029  <--  p - a0 + (p_tol + a0_tol)
>   8.90000000000008029  <--  p - a0 + p_tol + a0_tol    [NEW]
>   8.90000000000007851  <--  p + p_tol - a0 + a0_tol
> ```
> > Only the 3rd and 4th option is closest to (and occasionally exactly) the actual sum (and is largest). In the first enhancement step, I used the 4th option. 

#### 2.2 other changes

* in `region.CartesianGrid2D.get_index_of()`, I also [removed `numpy.array()`](https://github.com/SCECcode/pycsep/commit/9c14ee8) (redundant with first lines in `bin1d_vec`; should slightly speed up spatial binning)
* [drop any `tol=0.00001` from `bin1d_vec` calls or functions that call it](https://github.com/SCECcode/pycsep/commit/0b95738) (only used when binning magnitudes), even in the unit tests

<br>

> Notes (for those interested):
> * specifying `tol=0.00001` was not even necessary with the revised tolerance section by [commit 949d342c](https://github.com/SCECcode/pycsep/commit/949d342c) (albeit most of the `tol=0.00001` were introduced by this commit);
>    * Why? because magnitude bins are "cleanly" created with `np.arange()` in [`csep.core.regions.magnitude_bins`](https://docs.cseptesting.org/_modules/csep/core/regions.html#magnitude_bins);
>    * their redundancy is confirmed by passing the extended unit test `test_scalar_inside` (see below) without `tol=0.00001` using the unfixed `bin1d_vec`;
>    * yet, I left the optional `tol=None` argument in `bin1d_vec` and the functions that call it – just in case there is a real need to override the tolerance.

### 3. [Extend unit tests](https://github.com/SCECcode/pycsep/commit/92e33f0470da28edea8371e4687c954d5d8278b1)

* in `test_calc.TestBin1d`:
    * add `test_bin1d_single_bin2`, which inputs a single bin _and_ a single point;
    * split off a new `test_scalar_inside` from `test_scalar_outside`, which checks for (magnitude) bin edges `5.95, 6.05, ..., 8.95`:
        * _all_ corner cases (*.*5)
        * _all_ center bins (*.*0)
        * _all_ end of bins (*.*99999999)
        * one large value (10)
* add separate class `TestBin2d` in `test_region`, which makes more realistic unit tests by extending the self-check mentioned in #255 and performing it...
    * for three regions: Italy, California, NZ (New-Zealand);
    * for all origins, mid-points, and end-corners (in the bin at the opposite side of origin);
    * for double precision (_float64_ / _f8_) and single precision (_float32_ / _f4_) of the _points_ (not bins);
    * as loop (over each point individually) and single vector (all points at once).
    * ==> 36 unit test combinations
    * (albeit targeting `bin1d_vec`, it also unit-tests region's [`CartesianGrid2D.get_index_of`](https://github.com/SCECcode/pycsep/blob/2feb9148/csep/core/regions.py#L618) and by extension [`GriddedForecast.get_index_of`](https://docs.cseptesting.org/reference/generated/csep.core.forecasts.GriddedForecast.get_index_of.html))

<br>

> Notes (for those interested):
> * the _float32_-based unit tests only pass after the first enhancement in the previous section;
>     * (_float32_ is not a use case when using pyCSEP's implemented [readers](https://github.com/SCECcode/pycsep/blob/main/csep/utils/readers.py) [normally](https://docs.cseptesting.org/concepts/catalogs.html#id7) (read values will eventually be converted to `float64` [here](https://github.com/SCECcode/pycsep/blob/2feb9148/csep/core/catalogs.py#L284) by using [these specified dtypes](https://github.com/SCECcode/pycsep/blob/2feb9148/csep/core/catalogs.py#L883)), but it could happen with a custom loader function and the intention to save some memory - note that the [reader module](https://github.com/SCECcode/pycsep/blob/main/csep/utils/readers.py) still reveals some (incorrect) mentions of `float32`/`f4` in the docstrings). 
> * the unit tests for the NZ region only pass after addressing another issue in the next section.

### 4. Another issue: imprecise bin edge creation

For the NZ region, the end-corner-based unit test (`test_bin2d_regions_endcorner`) failed; inspecting the problematic corner point for the Longitude yields:

```
      origin: 167.79999999999998
           p: 167.89998999999997
          a0: 165.69999
           h: 0.09999999999999432
      p - a0: 2.19999999999996
```
It took me a while to spot the culprit; but this helped:
```
region.xs: [165.69999, 165.79999, 165.89999, 165.99999, 166.09999, 166.19999, 166.29999, 166.39999, ...]
dh: 0.1
```

Oops, they are not rounded to the first decimal digit despite `region.dh == 0.1`!

> FYI: The chain leading to failed test:
>  1. improperly rounded `.xs` (`.*9999`)
>  2. ... causes a likewise imprecise `a0`
>  3. ... which causes `(p - a0 + p_tol + a0_tol) / (h - h_tol)` to be just above `22` (`22.000099999901586`), instead of being just below (`21.99999999990184`)
>  4. ... which eventually nudges `idx` into the next bin

Apparently, the underlying [`csep.utils.calc.cleaner_range()`](https://github.com/SCECcode/pycsep/blob/2feb9148/csep/utils/calc.py#L204) is the culprit. It turns out that the particularly chosen `const = 100000` leads to weird numerical imprecision (it doesn't happen if it were 10'000 or 1'000'000):
```
165.7 * 100000  = 16569999.999999998 <-- the chosen `const`
165.7 * 10000   = 1657000.0
165.7 * 1000000 = 165700000.0
```
Then `np.floor(const * start)` operation yields `16569999.0` instead of `1657000.0`.

[Numpy's user guide](https://numpy.org/doc/stable/user/how-to-partition.html) suggest:
> _Use `numpy.linspace` if you want the endpoint to be included in the result, or if you are using a **non-integer step size**._

I tried it, but some bin edges still contained `.*99999...`.

Instead, I [modified `cleaner_range()` only in some other aspects](https://github.com/SCECcode/pycsep/commit/a836927). The crucial modification is using `round` instead of `floor` – very related to the [problematics that lead me to this PR](https://github.com/SCECcode/pycsep/issues/255#issuecomment-2548687127). The other change – replacing the hard-coded `const` with a flexible `scale` parameter – is supposed to account for the number of decimal digits of bin edges _and_ stepping (`const = 100000` would lead to imprecise bin edges if `h < 0.000'01` or if the bin edges start at `*.000'001` – maybe someone needs that in the future 😉).

> Note (for those interested):
> This sub-issue is apparently caused by using `numpy.loadtxt()` inside `nz_csep_region()` instead of [`parse_csep_template()`](https://github.com/SCECcode/pycsep/blob/2feb9148/csep/core/regions.py#L334) as used for reading the California and Italy CSEP region, which uses `float()` from the Python standard library. So this additional fix makes bin creation more robust to input produced with different reader functions.

In the [same commit](https://github.com/SCECcode/pycsep/commit/a836927), I also simplified `core.regions.magnitude_bins()` to just call (this improved) `cleaner_range()`; it was essentially a copy of it.

### 5. Implication on existing test results

As mentioned in my [original comment of issue #255](https://github.com/SCECcode/pycsep/issues/255), test results produced after this PR will become irreproducible with past results **_only if_** the test catalog contains one or more events whose coordinates align with the region's spatial bin edges, e.g., a coarse single-decimal-digit coordinate like 42.3° and a gridding of 0.1°. Otherwise, this issue and PR is irrelevant.

But even if a catalog _does_ contain such events, I don't expect test results would change significantly – perhaps only if _all_ events have coarse locations. Be reminded that I spotted this whole issue only due to a difference at the 3-5th decimal digit for IGPE compared to an independent binning implementation; the test catalog contained five of such events (262 events in total).

I still wanted to assess if this PR leads to some irreproducibility, so I ran our [first reproducibility package](https://github.com/wsavran/pycsep_esrl_reproducibility) using the most recent pyCSEP state (v0.6.3 + 22 commits) – once without and once with all changes in this PR. Eventually, they were exactly the same (up to the last [15th] decimal digit).[^1] Apparently, it doesn't involve events with coarse locations (this is a good thing!).

Apart that, I didn't do other comparisons; feel free to suggest me some (they should contain events with coarse locations and/or involve the NZ region).

---

Closes #255

[^1]: Btw: they were not exactly the same as the [expected output](https://github.com/wsavran/pycsep_esrl_reproducibility/tree/main/expected_output/results) due to some occasional mismatches at the last (15th) decimal digit. These slight mismatches are likely due to using a different platform/OS and/or a different combination of packages (I installed the most up-to-date versions of packages specified in pyCSEP's `requirements.yml` + adjusted `numpy==1.22.4` and `matplotlib==3.5.2` to get a compatible environment for the newest pyCSEP version).

### CODE REVIEW

**Comment by mherrmann3** [comment]
### Bonus: using `decimal` package

For the sake of curiosity, I also considered the `decimal` package as [suggested by Bill in the original issue](https://github.com/SCECcode/pycsep/issues/255#issuecomment-2515098482) (actually way before this PR and the aha-moment in [my comment](https://github.com/SCECcode/pycsep/issues/255#issuecomment-2548687127)).
Here necessary changes:

```diff
-    bins = numpy.array(bins)
-    p = numpy.array(p)
+    def array2decimal(x):
+        x = x.squeeze()  # omit superfluous dims
+        if x.ndim:
+            return np.array(
+                list(
+                    map(Decimal,
+                        numpy.array2string(
+                            x
+                            max_line_width=np.inf, threshold=np.inf
+                        )[1:-1].split())
+                    )
+            )
+        else:
+            return np.array(Decimal(numpy.array2string(x)))
+    
+    
+    p = array2decimal(numpy.array(p))
+    bins = array2decimal(numpy.array(bins))
```
and
```diff
-    # Deal with border cases
-    a0_tol = numpy.abs(a0) * numpy.finfo(numpy.float64).eps
-    h_tol = numpy.abs(h) * numpy.finfo(numpy.float64).eps
-    p_tol = numpy.abs(p) * numpy.finfo(numpy.float64).eps
-    
-    # absolute tolerance
-    if tol is None:
-        idx = numpy.floor((p + (p_tol + a0_tol) - a0) / (h - h_tol))
-    else:
-        idx = numpy.floor((p + (tol + a0_tol) - a0) / (h - h_tol))
-    if h < 0:
-        raise ValueError("grid spacing must be positive and monotonically increasing.")
-    # account for floating point uncertainties by considering extreme case
+    idx = (p - a0) / h
```

But the performance is abysmal due to mapping via strings (that's how the `decimals` package works to circumvent numerical precision):

| Enhancement                                       | **vectorized** |  looped |
|:--------------------------------------------------|---------------:|--------:|
| `decimal`                                         |    **17.1 ms** | 2.25 s  |

So 320x / 50x slower than the current implementation.

Also, it doesn't pass some unit tests.

Additionally, [array2string](https://numpy.org/doc/stable/reference/generated/numpy.array2string.html) must be well configured to avoid creating any weird strings from floats due to numerical precision (alternatively, one could represent the region's origins/bins directly as `Decimal`s on initializations, but that's all not necessary since we solved the numerical issues above).

**Comment by mjw98** [comment]
Wow, Marcus! Thank you!
________________________________

**Comment by wsavran** [comment]
@mherrmann3 The new solution looks much simpler and more optimal than rounding in the issue. Nice Work! I'm encouraged by your tests with the reproducibility package. I'm looking into why some of the builds are failing for different OS versions right now. It's failing due to an error in the `comcat` package but I haven't tracked down the reason for the `vcr` package throwing errors in some version but not others.

**Comment by wsavran** [comment]
it looks like it might be related to the pinned vcrpy installation we are using in the workflows/build-test.yml file 
```pip install vcrpy==4.3.1 pytest pytest-cov``` 

there was a warning in the release notes on vcrpy that this version might not play nicely with urllib3. the last successful builds used `urllib3==2.2.3` and the failing builds in this pr are using `urllib3==2.3.0`. 

it looks like we have some options here:
1) remove the pinned version from `.github/workflows/build-test.yml`
2) create a `requirements-test.yml` that includes sphinx and vcrpy (to mimic the `requirements-dev.txt`) and use this in the workflow
3) add vcrpy, pytest, and pytest-cov to the `create-args` directive under `jobs >> steps` in `.github/workflows/build-test.yml`.

@mherrmann3 do you have any suggestions or thoughts on this?

**Comment by mherrmann3** [comment]
Hey Bill, I'm not experienced with github's workflows, but I'd prefer the simplest option, number 1, in your list.

I don't see any reason for the pinning. It seems that @pabloitu simply pinned `vcrpy` to [`4.3.1`](https://github.com/kevin1024/vcrpy/releases/tag/v4.3.1) as it was the most recent version [back at that time](https://github.com/SCECcode/pycsep/commit/cb54bd04ca1721ac4a582e34b0d791e4d1109bbf) (and then [re-used this part also for `build_pip`](https://github.com/SCECcode/pycsep/commit/c887782a682cbed09106c49319e3b33a207cee66)).

So we could unpin it twice in `.github/workflows/build-test.yml`. You go ahead?

**Comment by wsavran** [comment]
@pabloitu Any thoughts on this?

**Comment by wsavran** [comment]
great news about the checks. i reviewed the code and it looks good to me. if someone else takes a look at it then we should be good to go. really nice and thorough job on this pr @mherrmann3!

**Review by pabloitu — CHANGES_REQUESTED** [review]
@mherrmann3 Awesome PR Marcus! Thanks for the detailed explanations and the literature to address the issue and propose a solid solution to it C:! 

I've tested this PR with my own results (e.g., Italy 2010) and there are changes up to 1e-12, only in test statistic distributions, which I guess are due to precision changes rather than EQs-cell assignment. I consider the changes to be negligible, in contrast to be sure now that the eq assignment to cells is now done thoroughly!

Could you do a last rebase on the latest commits and describe the `tol` argument in docstrings (you could do it much better than I do). It would help immensely when someone in the future modifies the functions, so they know what this argument does. I will approve and merge when you have the time to do this.

File: csep/core/catalogs.py [hunk 1]
```diff
@@ -699,7 +699,7 @@ def spatial_event_probability(self):
         event_flag[idx] = 1
         return event_flag
 
-    def magnitude_counts(self, mag_bins=None, tol=0.00001, retbins=False):
+    def magnitude_counts(self, mag_bins=None, tol=None, retbins=False):
```
[the remark this answers is not in the document]
Comment (mherrmann3) [reply csep/core/catalogs.py:702]: Done, see https://github.com/SCECcode/pycsep/pull/270/commits/d7f51bb6ed7a6b3a10f60cb5f1854d98bf6acb2c

File: tests/test_regions.py [hunk 2]
```diff
@@ -86,3 +86,72 @@ def test_origins(self):
...
+
+class TestBin2d(unittest.TestCase):
+
+    @classmethod
+    def setUpClass(cls):
+        super(TestBin2d, cls).setUpClass()
+
+        # (loading those is the bottleneck of this test case)
+        cls.regions = [
+            italy_csep_region(),
```
[the remark this answers is not in the document]
Comment (mherrmann3) [reply tests/test_regions.py:102]: Great idea! But you are right, `global_region` would take too much time (~2min loading + ~5min per loop + ~4s per vect), so I ignored it - for now (waiting for quantum computing 😉).

See: https://github.com/SCECcode/pycsep/pull/270/commits/f5ff69dc449111802086092b6ac2fb620b52452f

> Btw: For `nz_csep_collection_region` and `global_region`, the **original** (unfixed) binning implementation didn't fail the self-check loop (so the `bin1d_vec` fix doesn't affect those)
> * for `global_region` because it directly generates the origins with `cleaner_range`
> * for `nz_csep_collection_region`, I'm not sure why (didn't investigate it further)
<br>

> Btw: Albeit `global_region()` is not loading any files, `CartesianGrid2D()` causes a massive slow-down;
> specifically its `_build_bitmask_vec()`, which
>  1. builds `midpoints = ...` by calling `poly.centroid()` as many times as there are spatial bins
>  2. loops again over each bin to set the mask
> 
> The global region as 6.5 million bins vs. ~10k for the IT/Cal/NZ region
> So besides a slower loading, a more significant contribution to slower unit tests would be the self check _loops_.

File: csep/core/catalogs.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (pabloitu) [reply csep/core/catalogs.py:702]: Awesome

File: tests/test_regions.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (pabloitu) [reply tests/test_regions.py:102]: haha, awesome. Thanks

[the remark this answers is not in the document]
Comment (pabloitu) [reply tests/test_regions.py:102]: Ouch, the CI is failing with the latest magnitude-test-tests we merged. I will take a look at it to see why bin1d_vec is failing. Weirdly bin1d_vec is only failing in the new M=tests, as well as only failing in Unix, but passing in OSX nor Windows.

**Comment by mherrmann3** [comment]
I already provided a fix for the failed builds.

This happened because the unit tests of #268 featured a special case: a single bin at position 0, which was directly used (incorrectly) as arbitrary bin size. I introduced this incorrect behavior in this very PR with https://github.com/SCECcode/pycsep/pull/270/commits/29d3bb326fe88036de1691fef32a813acf62034a, which was apparently wrong and overambitious. I restored the behavior for this aspect.

Next time, I'll run all unit tests locally before pushing :-)

**Comment by pabloitu** [comment]
Great catch! Eventually we will need one-bin regions well-defined, where we are interested only in the temporal component (e.g., some fluid injection scenarios, etc).

**Review by pabloitu — APPROVED** [review]
